testing-guidelines/ci.md

```diff
@@ -0,0 +1,431 @@
+
+These changesets appear in the `.changeset` directory as markdown files and each changeset has its own dedicated file (this reduces conflicts). These changeset can be written manually or generated via a CLI but we try to automatically detect the changes in a PR and generate a changeset. This isn't always perfect but it is pretty reliable. The changeset can be updated by modifying labels, editing the issue, or modifying the file directly.
+
+#### merging into `main`
+
+When a pull request is merging to main there are two possible routes.
+
+- Create or update a versioning PR
+- Publish packages
+
+A normal pull request created by a user that does not contain a changeset will do nothing.
+
+A normal pull request created by a user that _does_ contain a changeset will either create or update a versioning PR. This is a PR that tracks the changelog and versions that will be used for the release. The changelogs for each package are generated from the descriptions in the changeset files. The generated version pull request is on a branch named `changeset-release/main` and there is only ever one (unless we are also working on a pre-release).
+
+When the `changeset-release/main` branch is merged into main all of the necessary versions will have been bumped and the changelogs generated. We then go through to see what needs to be published and publish to PyPi for Python packages and npm for javascript pacakges.
+
+
+## Architecture
+
+The CI on this repo is a little unconventional, this is mainly to work around various gaps in the GitHub Actions API while solving for our use case.
+
+The [technical details](#technical-details) below explain how things are setup, the [context](#context) sections explains what problem we are trying to solve with this architecture, why we took this approach, and the various challenges we faced along the way. It is a little 'extra' for repo docs but serves as good context for the future.
+
+### Technical details
+
+Most our our workflows are contained within the repository but we also have some custom JavaScript actions that are stored in the [gradio-app/github](https://github.com/gradio-app/github) repository. These actions will be mentioned below but I'll put a list here for ease of navigation:
+
+- [`find-pr`](https://github.com/gradio-app/github/tree/main/packages/find-pr) - for getting a pr number, pr branch, and other metadata for a `workflow_run`.
+- [`comment-pr`](https://github.com/gradio-app/github/tree/main/packages/comment-pr) - for creating and updating comments on a pull request.
+- [`commit-status`](https://github.com/gradio-app/github/tree/main/packages/commit-status) - for creating statuses on commits.
+- [`generate-changeset`](https://github.com/gradio-app/github/tree/main/packages/generate-changeset) - for generating and updating changesets based on the state or a pr and branch.
+- [`publish-pypi`](https://github.com/gradio-app/github/tree/main/packages/publish-pypi) - for publishing packages to pypi.
```

Comment:
I've had great success using pypa's own action for the actual publishing, no need to faff around with twine and secrets and so on (especially when using trusted publishers): https://github.com/marketplace/actions/pypi-publish

Reply:
We used to use it but its really slow because it is a docker action, so i wrote a js one that is much faster.

Reply:
It isn't really slow at all for me – it does depend on bits being cached of course...

Reply:
Its spins up a docker container, so its always going to be slower than a js action but each to their own.

Reply:
Oh the other thing is you have to use the action once per publish, so if you are publishing multiple packages, it isn't ideal.